### PR TITLE
[TF Backend] Wrap tf.where into Keras 

### PR DESCRIPTION
**Pull request by tomguluson92**
### Summary

Hi @fchollet, I wrap tf.where into Keras: K.where. An easy wrapper~

### PR Overview

- [y] This PR requires new unit tests [y/n] (make sure tests are included)
- [y] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [y] This PR is backwards compatible [y/n]
- [y] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -1828,6 +1828,25 @@ def less_equal(x, y):
...
+
+    # Arguments
+        condition: Tensor or variable.
+        x: Tensor or variable.
+        y: Tensor or variable.
+
+    # Returns
+        A tensor.
+    """
+    if x is not None and y is not None:
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: It seems like this `if` statement would have no effect?

Comment (tomguluson92) [reply keras/backend/tensorflow_backend.py]: Thx, I had it revised.

**Comment by tomguluson92** [comment]
@fchollet Thx a lot, i add it in all backend. But i had not found there is a `where` ops in CNTK which can performs same function like `tf.where`.

File: tests/keras/backend/backend_test.py [hunk 2]
```diff
@@ -159,6 +167,30 @@ def check_two_tensor_operation(function_name,
...
+                                 backend_list,
+                                 **kwargs):
+    x_shape, x_val = parse_shape_or_val(x_shape_or_val, "where")
+    y_shape, y_val = parse_shape_or_val(y_shape_or_val)
+    z_shape, z_val = parse_shape_or_val(z_shape_or_val)
+
+    t_list = []
+    z_list = []
+    for k in backend_list:
+        if k != KC:
```
Comment (gabrieldemarmiesse) [inline tests/keras/backend/backend_test.py]: `check_three_tensor_operation` is supposed to be generic. If you want to exclude cntk, you should do it before calling this function.

Comment (tomguluson92) [reply tests/keras/backend/backend_test.py]: ok, i have it fixed.

File: keras/backend/cntk_backend.py [hunk 3]
```diff
@@ -1111,6 +1111,10 @@ def less_equal(x, y):
     return C.less_equal(x, y)
 
 
+def where(condition, x, y):
+    raise NotImplementedError('CNTK not support ops like `tf.where`.')
```
Comment (fchollet) [inline keras/backend/cntk_backend.py]: @souptc is there any CNTK equivalent we could be using here? Thanks!

File: keras/backend/tensorflow_backend.py [hunk 4]
```diff
@@ -1828,6 +1828,24 @@ def less_equal(x, y):
     return tf.less_equal(x, y)
 
 
+def where(condition, x, y):
+    """Returns elements, either from x or y, depending on condition.
+       If only condition is given, \
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Don't indent. Don't use \ (this is not code)

File: keras/backend/tensorflow_backend.py [hunk 5]
```diff
@@ -1828,6 +1828,24 @@ def less_equal(x, y):
...
+       (suppose we have transfer condition from Tensor to `numpy.ndarray`)
+       Return `tf.convert_to_tensor(condition.nonzero())`.
+
+    # Arguments
+        condition: Tensor or variable.
+        x: Tensor or variable.
+        y: Tensor or variable.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Please use more informative descriptions, not just types

File: keras/backend/tensorflow_backend.py [hunk 6]
```diff
@@ -1828,6 +1828,24 @@ def less_equal(x, y):
...
+
+    # Returns
+        A tensor.
+    """
+    condition = tf.cast(condition, dtype=tf.bool)
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Make the cast conditional on `condition` being the wrong dtype

Comment (fchollet) [reply keras/backend/tensorflow_backend.py]: (is the cast really necessary? does the statement below crash without the cast?)

File: tests/keras/backend/backend_test.py [hunk 7]
```diff
@@ -59,11 +59,19 @@ def cntk_func_tensors(function_name, shapes_or_vals, **kwargs):
     return output_cntk, cntk_func
 
 
-def parse_shape_or_val(shape_or_val):
-    if isinstance(shape_or_val, np.ndarray):
-        return shape_or_val.shape, shape_or_val
-    else:
-        return shape_or_val, np.random.random(shape_or_val).astype(np.float32) - 0.5
+def parse_shape_or_val(shape_or_val, flags="normal"):
+    if flags == "normal":
```
Comment (fchollet) [inline tests/keras/backend/backend_test.py]: Use `'` as quote character for intra-file consistency

**Comment by gabrieldemarmiesse** [comment]
@taehoonlee maybe we could resurrect #11265 ? This would be useful for this PR, #11556 , and also the tests of the new K.clip function since clip can now take three input tensors as arguments.

To me, having `check_single_tensor_operation`, `check_two_tensor_operation`, `check_three_tensor_operation` isn't a good idea. It will be hard to maintain in the long run. Especially if we start making `check_five_tensor_operation` for batchnorm.

File: keras/backend/tensorflow_backend.py (hunk 6 above)
Comment (tomguluson92) [reply keras/backend/tensorflow_backend.py]: Yep, if i don't add this explicit cast,  The Travis CI build gonna failed.

File: keras/backend/cntk_backend.py (hunk 3 above)
Comment (souptc) [reply keras/backend/cntk_backend.py]: I don't think so, adding @KeDengMS from cntk team to confirm

Comment (KeDengMS) [reply keras/backend/cntk_backend.py]: You may use [element_select](https://cntk.ai/pythondocs/cntk.ops.html#cntk.ops.element_select)

Comment (tomguluson92) [reply keras/backend/cntk_backend.py]: @KeDengMS @souptc Thanks~

**Comment by gabrieldemarmiesse** [comment]
The tests are ok but we'll need to do a refactoring later. It is ok to merge.

**Comment by gabrieldemarmiesse** [comment]
@fchollet maybe a noob question, but what's the benefit of having K.where when we have K.switch? Isn't K.switch just a more powerful version of K.where which also handles callables?

**Comment by taehoonlee** [comment]
@fchollet, @gabrieldemarmiesse, I agree this point:
>>> To me, having check_single_tensor_operation, check_two_tensor_operation, check_three_tensor_operation isn't a good idea. It will be hard to maintain in the long run.

I think the following is enough to perform the unit test.
```python
def test_where(self):
    shape = (4, 2)
    cond = np.random.random(shape) > 0.5
    x = np.random.random(shape).astype(np.float32)
    y = np.random.random(shape).astype(np.float32)
    z = K.eval(K.where(K.variable(cond), K.variable(x), K.variable(y)))
    assert_allclose(z, KNP.where(cond, x, y), atol=1e-5)
```

**Comment by fchollet** [comment]
> maybe a noob question, but what's the benefit of having K.where when we have K.switch? Isn't K.switch just a more powerful version of K.where which also handles callables?

That's a very good point. `K.switch` is exactly `K.where`, except it also works for higher-dimensional cond arguments. I had not noticed this. As a result, I don't think we should merge this PR since it would make the backend API redundant.
